File: claude-code-sdk/src/main/kotlin/com/claudecodeplus/sdk/types/ToolTypes.kt
package com.claudecodeplus.sdk.types

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Contextual
import kotlinx.serialization.json.JsonElement

/**
 * 工具类型枚举
 *
 * 定义所有已知的Claude Code工具类型。
 * 这些类型与官方文档中的工具名称保持一致。
 */
enum class ToolType(val toolName: String) {
    // 基础文件操作工具
    BASH("Bash"),
    BASH_OUTPUT("BashOutput"),
    KILL_SHELL("KillShell"),
    EDIT("Edit"),
    MULTI_EDIT("MultiEdit"),
    READ("Read"),
    WRITE("Write"),

    // 搜索和查找工具
    GLOB("Glob"),
    GREP("Grep"),

    // 网络工具
    WEB_FETCH("WebFetch"),
    WEB_SEARCH("WebSearch"),

    // 开发和任务管理工具
    TODO_WRITE("TodoWrite"),
    TASK("Task"),
    EXIT_PLAN_MODE("ExitPlanMode"),

    // Jupyter notebook工具
    NOTEBOOK_EDIT("NotebookEdit"),

    // MCP (Model Context Protocol) 工具
    MCP_TOOL("mcp__"),  // MCP工具通常以mcp__开头
    LIST_MCP_RESOURCES("ListMcpResourcesTool"),
    READ_MCP_RESOURCE("ReadMcpResourceTool"),

    // 未知工具类型
    UNKNOWN("unknown");

    companion object {
        /**
         * 根据工具名称获取对应的工具类型
         */
        fun fromToolName(name: String): ToolType {
            return when {
                name.startsWith("mcp__") -> MCP_TOOL
                else -> values().find { it.toolName == name } ?: UNKNOWN
            }
        }
    }
}

/**
 * 具体工具使用的基础接口
 *
 * 这个接口继承自ContentBlock，确保向后兼容性。
 * 所有具体的工具类都应该实现这个接口。
 *
 * 注意：
 * - 每个子类使用独特的 @SerialName 用于 Kotlin 序列化的多态性
 * - name 字段包含工具名称(如 "TodoWrite", "Edit", "Write" 等)
 * - input 字段包含原始的工具参数 (与 Claude API 格式一致)
 * - 在序列化为 JSON 发送给前端时,应该手动构建 {type: "tool_use", name: "...", id: "...", input: {...}} 格式
 */
@Serializable
sealed interface SpecificToolUse : ContentBlock {
    val id: String              // 工具调用ID
    val name: String            // 工具名称，如 "TodoWrite", "Edit", "Write" 等
    val input: JsonElement      // 原始参数 (与 Claude API 格式一致)

    // 内部使用的枚举类型，不序列化到 JSON
    @kotlinx.serialization.Transient
    val toolType: ToolType

    /**
     * 获取强类型的参数
     * 子类应该实现这个方法来提供特定工具的参数访问
     * 注意：返回类型为Any，但实际运行时类型是已知的
     */
    fun getTypedParameters(): Map<String, @Contextual Any>
}

/**
 * Bash工具使用
 * 执行shell命令
 */
@Serializable
@SerialName("bash_tool_use")
data class BashToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val command: String,
    val description: String? = null,
    val timeout: Long? = null,
    val runInBackground: Boolean = false
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.BASH

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("command", command)
        description?.let { put("description", it) }
        timeout?.let { put("timeout", it) }
        put("run_in_background", runInBackground)
    }
}

/**
 * Edit工具使用
 * 编辑文件内容
 */
@Serializable
@SerialName("edit_tool_use")
data class EditToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val filePath: String,
    val oldString: String,
    val newString: String,
    val replaceAll: Boolean = false
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.EDIT

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "file_path" to filePath,
        "old_string" to oldString,
        "new_string" to newString,
        "replace_all" to replaceAll
    )
}

/**
 * MultiEdit工具使用
 * 对单个文件进行多次编辑
 */
@Serializable
@SerialName("multi_edit_tool_use")
data class MultiEditToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val filePath: String,
    val edits: List<EditOperation>
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.MULTI_EDIT

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "file_path" to filePath,
        "edits" to edits.map { edit ->
            mapOf(
                "old_string" to edit.oldString,
                "new_string" to edit.newString,
                "replace_all" to edit.replaceAll
            )
        }
    )

    @Serializable
    data class EditOperation(
        val oldString: String,
        val newString: String,
        val replaceAll: Boolean = false
    )
}

/**
 * Read工具使用
 * 读取文件内容
 */
@Serializable
@SerialName("read_tool_use")
data class ReadToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val filePath: String,
    val offset: Int? = null,
    val limit: Int? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.READ

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("file_path", filePath)
        offset?.let { put("offset", it) }
        limit?.let { put("limit", it) }
    }
}

/**
 * Write工具使用
 * 写入文件内容
 */
@Serializable
@SerialName("write_tool_use")
data class WriteToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val filePath: String,
    val content: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.WRITE

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "file_path" to filePath,
        "content" to content
    )
}

/**
 * Glob工具使用
 * 文件模式匹配
 */
@Serializable
@SerialName("glob_tool_use")
data class GlobToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val pattern: String,
    val path: String? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.GLOB

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("pattern", pattern)
        path?.let { put("path", it) }
    }
}

/**
 * Grep工具使用
 * 文本搜索
 */
@Serializable
@SerialName("grep_tool_use")
data class GrepToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val pattern: String,
    val path: String? = null,
    val outputMode: String? = null,
    val glob: String? = null,
    val type: String? = null,
    val caseInsensitive: Boolean = false,
    val showLineNumbers: Boolean = false,
    val contextBefore: Int? = null,
    val contextAfter: Int? = null,
    val headLimit: Int? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.GREP

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("pattern", pattern)
        path?.let { put("path", it) }
        outputMode?.let { put("output_mode", it) }
        glob?.let { put("glob", it) }
        type?.let { put("type", it) }
        put("-i", caseInsensitive)
        put("-n", showLineNumbers)
        contextBefore?.let { put("-B", it) }
        contextAfter?.let { put("-A", it) }
        headLimit?.let { put("head_limit", it) }
    }
}

/**
 * WebFetch工具使用
 * 获取网页内容
 */
@Serializable
@SerialName("web_fetch_tool_use")
data class WebFetchToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val url: String,
    val prompt: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.WEB_FETCH

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "url" to url,
        "prompt" to prompt
    )
}

/**
 * WebSearch工具使用
 * 网络搜索
 */
@Serializable
@SerialName("web_search_tool_use")
data class WebSearchToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val query: String,
    val allowedDomains: List<String>? = null,
    val blockedDomains: List<String>? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.WEB_SEARCH

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("query", query)
        allowedDomains?.let { put("allowed_domains", it) }
        blockedDomains?.let { put("blocked_domains", it) }
    }
}

/**
 * TodoWrite工具使用
 * 管理待办事项列表
 */
@Serializable
@SerialName("todo_write_tool_use")
data class TodoWriteToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val todos: List<TodoItem>
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.TODO_WRITE

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "todos" to todos.map { todo ->
            mapOf(
                "content" to todo.content,
                "status" to todo.status,
                "activeForm" to todo.activeForm
            )
        }
    )

    @Serializable
    data class TodoItem(
        val content: String,
        val status: String,  // pending, in_progress, completed
        val activeForm: String
    )
}

/**
 * Task工具使用
 * 启动子任务代理
 */
@Serializable
@SerialName("task_tool_use")
data class TaskToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val description: String,
    val prompt: String,
    val subagentType: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.TASK

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "description" to description,
        "prompt" to prompt,
        "subagent_type" to subagentType
    )
}

/**
 * NotebookEdit工具使用
 * 编辑Jupyter Notebook
 */
@Serializable
@SerialName("notebook_edit_tool_use")
data class NotebookEditToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val notebookPath: String,
    val newSource: String,
    val cellId: String? = null,
    val cellType: String? = null,
    val editMode: String? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.NOTEBOOK_EDIT

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("notebook_path", notebookPath)
        put("new_source", newSource)
        cellId?.let { put("cell_id", it) }
        cellType?.let { put("cell_type", it) }
        editMode?.let { put("edit_mode", it) }
    }
}

/**
 * MCP工具使用
 * Model Context Protocol工具的通用容器
 */
@Serializable
@SerialName("mcp_tool_use")
data class McpToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val fullToolName: String,  // 完整的工具名称，如 "mcp__server_name__function_name"
    val serverName: String,
    val functionName: String,
    val parameters: Map<String, @Contextual Any>
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.MCP_TOOL

    override fun getTypedParameters(): Map<String, Any> = parameters
}

/**
 * BashOutput工具使用
 * 获取Bash命令的输出
 */
@Serializable
@SerialName("bash_output_tool_use")
data class BashOutputToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val bashId: String,
    val filter: String? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.BASH_OUTPUT

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        put("bash_id", bashId)
        filter?.let { put("filter", it) }
    }
}

/**
 * KillShell工具使用
 * 终止运行中的Shell进程
 */
@Serializable
@SerialName("kill_shell_tool_use")
data class KillShellToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val shellId: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.KILL_SHELL

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "shell_id" to shellId
    )
}

/**
 * ExitPlanMode工具使用
 * 退出计划模式
 */
@Serializable
@SerialName("exit_plan_mode_tool_use")
data class ExitPlanModeToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val plan: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.EXIT_PLAN_MODE

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "plan" to plan
    )
}

/**
 * ListMcpResources工具使用
 * 列出MCP服务器的可用资源
 */
@Serializable
@SerialName("list_mcp_resources_tool_use")
data class ListMcpResourcesToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val server: String? = null
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.LIST_MCP_RESOURCES

    override fun getTypedParameters(): Map<String, @Contextual Any> = buildMap {
        server?.let { put("server", it) }
    }
}

/**
 * ReadMcpResource工具使用
 * 读取指定的MCP资源
 */
@Serializable
@SerialName("read_mcp_resource_tool_use")
data class ReadMcpResourceToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val server: String,
    val uri: String
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.READ_MCP_RESOURCE

    override fun getTypedParameters(): Map<String, @Contextual Any> = mapOf(
        "server" to server,
        "uri" to uri
    )
}

/**
 * 未知工具使用
 * 用于处理不认识的工具类型
 */
@Serializable
@SerialName("unknown_tool_use")
data class UnknownToolUse(
    override val id: String,
    override val name: String,
    override val input: JsonElement,
    val toolName: String,
    val parameters: Map<String, @Contextual Any>
) : SpecificToolUse {
    @kotlinx.serialization.Transient
    override val toolType = ToolType.UNKNOWN

    override fun getTypedParameters(): Map<String, Any> = parameters
}

File: vscode-extension/src/sdk/claude/types/toolTypes.ts
/**
 * Tool types for Claude Agent SDK
 * Translated from Kotlin: claude-agent-sdk/src/main/kotlin/com/asakii/claude/agent/sdk/types/ToolTypes.kt
 */

import type { JsonValue, JsonObject } from './common';
import type { ToolUseLike, ContentBlock } from './contentBlocks';

/**
 * Tool type enum.
 *
 * Defines all known Claude Code tool types.
 */
export const ToolType = {
  // Basic file operation tools
  BASH: { toolName: 'Bash', type: 'CLAUDE_BASH' },
  BASH_OUTPUT: { toolName: 'BashOutput', type: 'CLAUDE_BASH_OUTPUT' },
  KILL_SHELL: { toolName: 'KillShell', type: 'CLAUDE_KILL_SHELL' },
  EDIT: { toolName: 'Edit', type: 'CLAUDE_EDIT' },
  MULTI_EDIT: { toolName: 'MultiEdit', type: 'CLAUDE_MULTI_EDIT' },
  READ: { toolName: 'Read', type: 'CLAUDE_READ' },
  WRITE: { toolName: 'Write', type: 'CLAUDE_WRITE' },

  // Search and find tools
  GLOB: { toolName: 'Glob', type: 'CLAUDE_GLOB' },
  GREP: { toolName: 'Grep', type: 'CLAUDE_GREP' },

  // Network tools
  WEB_FETCH: { toolName: 'WebFetch', type: 'CLAUDE_WEB_FETCH' },
  WEB_SEARCH: { toolName: 'WebSearch', type: 'CLAUDE_WEB_SEARCH' },

  // Development and task management tools
  TODO_WRITE: { toolName: 'TodoWrite', type: 'CLAUDE_TODO_WRITE' },
  TASK: { toolName: 'Task', type: 'CLAUDE_TASK' },
  EXIT_PLAN_MODE: { toolName: 'ExitPlanMode', type: 'CLAUDE_EXIT_PLAN_MODE' },
  ASK_USER_QUESTION: { toolName: 'AskUserQuestion', type: 'CLAUDE_ASK_USER_QUESTION' },
  SKILL: { toolName: 'Skill', type: 'CLAUDE_SKILL' },
  SLASH_COMMAND: { toolName: 'SlashCommand', type: 'CLAUDE_SLASH_COMMAND' },
  ENTER_PLAN_MODE: { toolName: 'EnterPlanMode', type: 'CLAUDE_ENTER_PLAN_MODE' },

  // Jupyter notebook tool
  NOTEBOOK_EDIT: { toolName: 'NotebookEdit', type: 'CLAUDE_NOTEBOOK_EDIT' },

  // MCP (Model Context Protocol) tools
  MCP_TOOL: { toolName: 'mcp__', type: 'MCP' },
  LIST_MCP_RESOURCES: { toolName: 'ListMcpResourcesTool', type: 'CLAUDE_LIST_MCP_RESOURCES' },
  READ_MCP_RESOURCE: { toolName: 'ReadMcpResourceTool', type: 'CLAUDE_READ_MCP_RESOURCE' },

  // Unknown tool type
  UNKNOWN: { toolName: 'unknown', type: 'UNKNOWN' },
} as const;

export type ToolTypeInfo = (typeof ToolType)[keyof typeof ToolType];
export type ToolTypeName = keyof typeof ToolType;

/**
 * Get tool type from tool name.
 */
export function getToolType(toolName: string): ToolTypeInfo {
  if (toolName.startsWith('mcp__')) {
    return ToolType.MCP_TOOL;
  }
  const entry = Object.values(ToolType).find((t) => t.toolName === toolName);
  return entry ?? ToolType.UNKNOWN;
}

// ============================================================================
// Specific Tool Use Types
// ============================================================================

/**
 * Base interface for specific tool use.
 *
 * This interface inherits from ContentBlock to ensure backward compatibility.
 * All specific tool classes should implement this interface.
 *
 * Note:
 * - name field contains tool name (like "TodoWrite", "Edit", "Write", etc.)
 * - input field contains raw tool parameters (matches Claude API format)
 * - When serializing to JSON for frontend, should manually construct {type: "tool_use", name: "...", id: "...", input: {...}} format
 */
export interface SpecificToolUse extends ToolUseLike {
  /** Tool call ID */
  id: string;
  /** Tool name, like "TodoWrite", "Edit", "Write", etc. */
  name: string;
  /** Raw parameters (matches Claude API format) */
  input: JsonValue;
  /** Internal enum type, not serialized to JSON */
  toolType: ToolTypeInfo;
}

/**
 * Bash tool use - execute shell commands.
 */
export interface BashToolUse extends SpecificToolUse {
  name: 'Bash';
  toolType: typeof ToolType.BASH;
  command: string;
  description?: string;
  timeout?: number;
  runInBackground?: boolean;
}

/**
 * Edit tool use - edit file content.
 */
export interface EditToolUse extends SpecificToolUse {
  name: 'Edit';
  toolType: typeof ToolType.EDIT;
  filePath: string;
  oldString: string;
  newString: string;
  replaceAll?: boolean;
}

/**
 * Edit operation for MultiEdit.
 */
export interface EditOperation {
  oldString: string;
  newString: string;
  replaceAll?: boolean;
}

/**
 * MultiEdit tool use - multiple edits on a single file.
 */
export interface MultiEditToolUse extends SpecificToolUse {
  name: 'MultiEdit';
  toolType: typeof ToolType.MULTI_EDIT;
  filePath: string;
  edits: EditOperation[];
}

/**
 * Read tool use - read file content.
 */
export interface ReadToolUse extends SpecificToolUse {
  name: 'Read';
  toolType: typeof ToolType.READ;
  filePath: string;
  offset?: number;
  limit?: number;
}

/**
 * Write tool use - write file content.
 */
export interface WriteToolUse extends SpecificToolUse {
  name: 'Write';
  toolType: typeof ToolType.WRITE;
  filePath: string;
  content: string;
}

/**
 * Glob tool use - file pattern matching.
 */
export interface GlobToolUse extends SpecificToolUse {
  name: 'Glob';
  toolType: typeof ToolType.GLOB;
  pattern: string;
  path?: string;
}

/**
 * Grep tool use - text search.
 */
export interface GrepToolUse extends SpecificToolUse {
  name: 'Grep';
  toolType: typeof ToolType.GREP;
  pattern: string;
  path?: string;
  outputMode?: string;
  glob?: string;
  type?: string;
  caseInsensitive?: boolean;
  showLineNumbers?: boolean;
  contextBefore?: number;
  contextAfter?: number;
  headLimit?: number;
}

/**
 * WebFetch tool use - fetch web page content.
 */
export interface WebFetchToolUse extends SpecificToolUse {
  name: 'WebFetch';
  toolType: typeof ToolType.WEB_FETCH;
  url: string;
  prompt: string;
}

/**
 * WebSearch tool use - web search.
 */
export interface WebSearchToolUse extends SpecificToolUse {
  name: 'WebSearch';
  toolType: typeof ToolType.WEB_SEARCH;
  query: string;
  allowedDomains?: string[];
  blockedDomains?: string[];
}

/**
 * Todo item for TodoWrite.
 */
export interface TodoItem {
  content: string;
  /** Status: pending, in_progress, completed */
  status: 'pending' | 'in_progress' | 'completed';
  activeForm: string;
}

/**
 * TodoWrite tool use - manage todo list.
 */
export interface TodoWriteToolUse extends SpecificToolUse {
  name: 'TodoWrite';
  toolType: typeof ToolType.TODO_WRITE;
  todos: TodoItem[];
}

/**
 * Task tool use - start subtask agent.
 */
export interface TaskToolUse extends SpecificToolUse {
  name: 'Task';
  toolType: typeof ToolType.TASK;
  description: string;
  prompt: string;
  subagentType: string;
}

/**
 * NotebookEdit tool use - edit Jupyter Notebook.
 */
export interface NotebookEditToolUse extends SpecificToolUse {
  name: 'NotebookEdit';
  toolType: typeof ToolType.NOTEBOOK_EDIT;
  notebookPath: string;
  newSource: string;
  cellId?: string;
  cellType?: string;
  editMode?: string;
}

/**
 * MCP tool use - Model Context Protocol tool generic container.
 */
export interface McpToolUse extends SpecificToolUse {
  toolType: typeof ToolType.MCP_TOOL;
  /** Full tool name, like "mcp__server_name__function_name" */
  fullToolName: string;
  serverName: string;
  functionName: string;
  parameters: JsonObject;
}

/**
 * BashOutput tool use - get Bash command output.
 */
export interface BashOutputToolUse extends SpecificToolUse {
  name: 'BashOutput';
  toolType: typeof ToolType.BASH_OUTPUT;
  bashId: string;
  filter?: string;
}

/**
 * KillShell tool use - terminate running shell process.
 */
export interface KillShellToolUse extends SpecificToolUse {
  name: 'KillShell';
  toolType: typeof ToolType.KILL_SHELL;
  shellId: string;
}

/**
 * ExitPlanMode tool use - exit plan mode.
 */
export interface ExitPlanModeToolUse extends SpecificToolUse {
  name: 'ExitPlanMode';
  toolType: typeof ToolType.EXIT_PLAN_MODE;
  plan: string;
}

/**
 * ListMcpResources tool use - list available MCP server resources.
 */
export interface ListMcpResourcesToolUse extends SpecificToolUse {
  name: 'ListMcpResourcesTool';
  toolType: typeof ToolType.LIST_MCP_RESOURCES;
  server?: string;
}

/**
 * ReadMcpResource tool use - read specified MCP resource.
 */
export interface ReadMcpResourceToolUse extends SpecificToolUse {
  name: 'ReadMcpResourceTool';
  toolType: typeof ToolType.READ_MCP_RESOURCE;
  server: string;
  uri: string;
}

/**
 * Skill tool use - call registered skill (like /codex, /commit, etc.).
 */
export interface SkillToolUse extends SpecificToolUse {
  name: 'Skill';
  toolType: typeof ToolType.SKILL;
  /** Skill name, like "codex" */
  skill: string;
  /** Optional arguments */
  args?: string;
}

/**
 * Unknown tool use - for handling unrecognized tool types.
 */
export interface UnknownToolUse extends SpecificToolUse {
  toolType: typeof ToolType.UNKNOWN;
  toolName: string;
  parameters: JsonObject;
}

/**
 * Union type for all specific tool uses.
 */
export type AnyToolUse =
  | BashToolUse
  | EditToolUse
  | MultiEditToolUse
  | ReadToolUse
  | WriteToolUse
  | GlobToolUse
  | GrepToolUse
  | WebFetchToolUse
  | WebSearchToolUse
  | TodoWriteToolUse
  | TaskToolUse
  | NotebookEditToolUse
  | McpToolUse
  | BashOutputToolUse
  | KillShellToolUse
  | ExitPlanModeToolUse
  | ListMcpResourcesToolUse
  | ReadMcpResourceToolUse
  | SkillToolUse
  | UnknownToolUse;
